File: tree/trie/trie2.cpp
#include <unordered_map>

using namespace std;

/*
 * 子节点用 unordered_map 实现
 */

struct TrieNode
{
    bool terminal;
    unordered_map<char, TrieNode*> children;
    TrieNode()
    {
        terminal = false;
        children = unordered_map<char, TrieNode*>();
    }
    ~TrieNode(){}
};

class Trie
{
public:
    Trie()
    {
        root = new TrieNode();
    }

    ~Trie()
    {
        if(root)
            delete_sub_tree(root);
        delete root;
        root = nullptr;
    }

    void insert(const string& word)
    {
        TrieNode *iter = root;
        for(const char& ch: word)
        {
            if((iter -> children).count(ch) == 0)
                (iter -> children)[ch] = new TrieNode();
            iter = (iter -> children)[ch];
        }
        iter -> terminal = true;
    }

    bool search(string word)
    {
        const TrieNode* iter = find(word);
        return iter && iter -> terminal;
    }

    bool startsWith(string prefix)
    {
        return find(prefix) != nullptr;
    }

private:
    TrieNode *root;

    void delete_sub_tree(TrieNode* node)
    {
        auto it = (node -> children).begin();
        while(it != (node -> children).end())
        {
            TrieNode *child = it -> second;
            if(child)
                delete_sub_tree(child);
            delete child;
            child = nullptr;
            ++it;
        }
    }

    const TrieNode* find(const string& prefix) const
    {
        TrieNode* iter = root;
        for(const char c: prefix)
        {
            if((iter -> children).count(c) == 0)
                return nullptr;
            iter = (iter -> children)[c];
        }
        return iter;
    }
};

File: cpp/tree/trie/trie2.py
"""
子节点用 dict 实现
"""

class TrieNode:
    def __init__(self):
        self.terminal = False
        self.children = dict()

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node_iter = self.root
        for c in word:
            if c not in node_iter.children:
                node_iter.children[c] = TrieNode()
            node_iter = node_iter.children[c]
        node_iter.terminal = True

    def search(self, word: str) -> bool:
        node_iter = self._find(word)
        return node_iter is not None and node_iter.terminal

    def startsWith(self, prefix: str) -> bool:
        return self._find(prefix) is not None

    def _find(self, prefix):
        node_iter = self.root
        for c in prefix:
            if c not in node_iter.children:
                return None
            node_iter = (node_iter.children)[c]
        return node_iter
